### ISSUE TITLE
[Investigation] Is it possible to use navigator.clipboard.write to auto-paste text/image?

### ISSUE THREAD

**Issue by Jack-Works** [DimensionDev/Maskbook#3131]
raised in #2990

**Comment by manishoo** [comment]
I've tried this and it's totally possible with a mandatory "clipboardRead" permission. I'm now investigating whether I can get that permission on demand

**Comment by manishoo** [comment]
- It's possible to write images and text to the clipboard using `navigator.clipboard.write`
- Texts can be pasted just fine. Because they use [this line](https://github.com/DimensionDev/Maskbook/blob/152493e1dcecd7a78a2c4127c4e951960819146b/packages/maskbook/src/extension/injected-script/addEventListener.ts#L124) rather than `dispatchEventRaw`
- But pasting Images uses `dispatchEventRaw` which doesn't seem to work. I suspect this is because of a lack of a `paste` event listener on any of the DOM elements. My workaround was to use `document.execCommand("paste")` which works as expected if the `clipboardRead` permission is granted.
- I'm now trying to get that permission on demand

**Comment by Jack-Works** [comment]
Oh I understand why `document.execCommand("paste")` requires `clipboardRead`

**Comment by manishoo** [comment]
So @Jack-Works do you generally agree about using `document.execCommand("paste")` and getting the permission on demand? Or is the `clipboardRead` permission is too sensitive?

If you do agree, **when do you think we should get the permission?** 
Some options are:
1. When user first connects to Minds.com from the dashboard (`SocialNetwork.Base.declarativePermissions`)
2. When user clicks on the imagePayload toggle in PostDialog
3. When user clicks finish button in PostDialog

In the last two options, which I think are more appropriate, I tried getting the permission using the `DashboardRoute.RequestPermission` page, since `browser.permissions` is not available in those contexts. But I'm now trying to figure out how to check whether the permission is already granted or not. Any ideas is appreciated

**Comment by Jack-Works** [comment]
Is the permission comes from `navigator.permissions` (available for any web sites) or by `browser.permissions`?

**Comment by manishoo** [comment]
This is how it is now:

![output](https://user-images.githubusercontent.com/16239413/117841082-f710af80-b291-11eb-9168-36733f7a249d.gif)

**Comment by Jack-Works** [comment]
Let me rephrase. If permission is required by `navigator.*` API, that is not acceptable because it is shared with the current SNS domain.
So `browser.permission.request` is good. I think we should approve the UI for this, popup when creating a new post is a bit surprising. Maybe we can add a new button in the composition dialog to indicate that we need permission for auto pasting (cc @OiCkilL?).

I have another concern is that we use a Web Extension polyfill on iOS, and this feature might be not available to polyfill. @jk234ert Is it possible for the native side to emulate "paste" in the webview?

**Comment by manishoo** [comment]
The solution I suggested may not be optimal because somebody may not want to grant permission in the first place. It wouldn't make sense for them to click the "grant clipboard permission" button. I think some additional alternative solutions may include:
- an "enable auto-paste" button in the Mask composer
- a toast message that prompts the user to enable auto paste, just like the `autoPasteFailed` toast message. It could say something like "the encrypted image has been copied to the clipboard, you can now paste it in the composer. Click here to enable auto-paste to automatically do that"
- the regular `autoPasteFailed` toast plus an "enable auto-paste" button in the `AutoPasteFailedDialog`
- clicking the "finish" button directly opens the `RequestPermission` page in dashboard.

These are just some ideas

**Comment by Jack-Works** [comment]
> "the encrypted image has been copied to the clipboard, you can now paste it in the composer. Click here to enable auto-paste to automatically do that"

This looks good to me but I don't know how others think

**Comment by Jack-Works** [comment]
Recently I add a new kind of entry for things like permission requesting (#3158). I'll move permission requesting to that place.

**Comment by Tedko** [comment]
Guys. Really good job!

**Comment by manishoo** [comment]
I've updated the code.
@Jack-Works I think your first idea was probably the simplest implementation.
> Maybe we can add a new button in the composition dialog to indicate that we need permission for auto pasting

So I implemented it until we there's a better idea:
![output](https://user-images.githubusercontent.com/16239413/118358072-6b00cf80-b592-11eb-97a9-5a998b3c3135.gif)

> Recently I add a new kind of entry for things like permission requesting (#3158). I'll move permission requesting to that place.

Let me know if I can help. I checked the PR out but couldn't figure out how to request the permission the new way
